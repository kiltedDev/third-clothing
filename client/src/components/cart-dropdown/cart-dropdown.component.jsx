import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { withRouter } from 'react-router-dom';

import { toggleCartHidden } from '../../redux/cart/cart.actions';
import { selectCartItems }  from '../../redux/cart/cart.selectors';
import CustomButton from '../custom-button/custom-button.component';
import CartItem from '../cart-item/cart-item.component';

import {
  CartDropdownContainer,
  CartDropdownItems,
  EmptyMessage
} from './cart-dropdown.styles'

const CartDropdown = ({ cartItems, history, dispatch }) => (
  <CartDropdownContainer className='cart-dropdown'>
    <CartDropdownItems
      className='cart-items'
      children={
        cartItems.length
        ? cartItems.map( cartItem => (
          <CartItem
            key={ cartItem.id}
            item={ cartItem }
          />
        )) : (
          <EmptyMessage
            className='empty-message'
            children='Your cart is empty'
          />
        )
      }
    />
    <CustomButton
      onClick={() => {
        history.push( '/checkout' );
        dispatch( toggleCartHidden() );
      }}
      children='GO TO CHECKOUT'
    />
  </CartDropdownContainer>
);

const mapStateToProps = createStructuredSelector({
  cartItems: selectCartItems
})

export default withRouter(connect(mapStateToProps)(CartDropdown));
