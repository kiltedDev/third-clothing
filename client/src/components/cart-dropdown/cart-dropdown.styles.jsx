import styled from 'styled-components';
import CustomButton from '../custom-button/custom-button.component';

export const CartDropdownContainer = styled.div`
  position: absolute;
  width: 240px;
  height: 340px;
  display: flex;
  flex-direction: column;
  padding: 20px;
  border: 1px solid black;
  background-color: white;
  top: 90px;
  right: 40px;
  z-index: 5;

  @media screen and (max-width: 800px) {
    width: calc(100vw - 20px);
    right: 0px;
    height: 80vh;
    top:70px;
  }
`

export const CartDropdownItems = styled.div`
  height: 240px;
  display: flex;
  flex-direction: column;
  @media screen and (max-width: 800px) {
    height: 75vh;
  }
`

export const EmptyMessage = styled.span`
  font-size: 18px;
  margin: 50px auto;
`

export const CartDropdownButton = styled(CustomButton)`
  margin-top: auto;
`

CartDropdownButton.displayName = 'CartDropdownButton';
EmptyMessage.displayName = 'EmptyMessage';
