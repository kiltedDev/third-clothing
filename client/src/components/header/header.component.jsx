import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import CartIcon from '../cart-icon/cart-icon.component';
import CartDropdown from '../cart-dropdown/cart-dropdown.component';
import { selectCartHidden } from '../../redux/cart/cart.selectors';
import { selectCurrentUser } from '../../redux/user/user.selectors';
import { signOutStart } from '../../redux/user/user.actions';

import { ReactComponent as Logo } from '../../assets/horn.svg';

import {
  HeaderContainer,
  LogoContainer,
  OptionsContainer,
  OptionLink
} from './header.styles';

const Header = ({ currentUser, hidden, signOutStart }) => (

  <HeaderContainer  className='header'>
    <LogoContainer className='logo-container' to='/'>
      <Logo className='logo' alt='horn of odin produced by flaticon.com' label='horn of odin produced by flaticon.com'/>
    </LogoContainer>
    <OptionsContainer>
      <OptionLink
        to='/shop'
        name='Shop'
        children='Shop'
      />
      <OptionLink
        to='/contact'
        name='Contact'
        children='Contact'
      />
      {
        currentUser ?
        (
          <OptionLink
            as='div'
            onClick={ signOutStart }
            name='Sign Out'
            children='Sign Out'
          />
        ) : (
          <OptionLink
            to='/signin'
            name='Sign In'
            children='Sign In'
          />
        )
      }
      <CartIcon />
    </OptionsContainer>
    { hidden ? null : <CartDropdown /> }
  </HeaderContainer>
)

const mapStateToProps = createStructuredSelector({
  currentUser: selectCurrentUser,
  hidden: selectCartHidden
});

const mapDispatchToProps = dispatch => ({
  signOutStart: () => dispatch( signOutStart() )
});

export default connect( mapStateToProps, mapDispatchToProps )(Header);
