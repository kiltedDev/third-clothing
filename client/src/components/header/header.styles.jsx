import styled, { css } from 'styled-components';
import { Link } from 'react-router-dom';

export const HeaderContainer = styled.div`
  height: 90px;
  width: 100vw;
  display: flex;
  justify-content: space-between;
  margin-bottom: 25px;
  position: fixed;
  top:0;
  left:0;
  padding: 10px 40px;
  background-color: white;
  z-index: 1;

  @media screen and (max-width: 800px) {
    padding: 10px 20px;
    margin-bottom: 20px;
  }
`

export const LogoContainer = styled( Link )`
  height: 70px;
  width: 70px;
  position:relative;
  top: 50%;
  transform: translateY(-50%);

  @media screen and (max-width: 800px) {
    width: 50px;
    height: 50px;
  }
`

export const OptionsContainer = styled.div`
  width: 50%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: flex-end;
  @media screen and (max-width: 800px) {
    width: 80%;
  }
`

export const OptionStyles = css`
  padding: 10px 15px;
  text-transform: uppercase;
  cursor: pointer;
  @media screen and ( max-width: 360px) {
    padding: 10px;
  }
`

export const OptionLink = styled(Link)`
  ${OptionStyles}
`
