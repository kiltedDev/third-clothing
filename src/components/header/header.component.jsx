import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import { auth } from '../../firebase/firebase.utils';
import CartIcon from '../cart-icon/cart-icon.component';
import CartDropdown from '../cart-dropdown/cart-dropdown.component';
import { selectCartHidden } from '../../redux/cart/cart.selectors';
import { selectCurrentUser } from '../../redux/user/user.selectors';

import { ReactComponent as Logo } from '../../assets/horn.svg';

import {
  HeaderContainer,
  LogoContainer,
  OptionsContainer,
  OptionLink
} from './header.styles';

const Header = ({ currentUser, hidden }) => (

  <HeaderContainer  className='header'>
    <LogoContainer className='logo-container' to='/'>
      <Logo className='logo' alt='horn of odin produced by flaticon.com' label='horn of odin produced by flaticon.com'/>
    </LogoContainer>
    <OptionsContainer>
      <OptionLink
        to='/shop'
        children='Shop'
      />
      <OptionLink
        to='/shop'
        children='Contact'
      />
      {
        currentUser ?
        (
          <OptionLink
            as='div'
            onClick={ () => auth.signOut() }
            children='Sign Out'
          />
        ) : (
          <OptionLink
            to='/signin'
            children='Sign In'
          />
        )
      }
      <CartIcon />
    </OptionsContainer>
    { hidden ? null : <CartDropdown /> }
  </HeaderContainer>
)

const mapStateToProps = createStructuredSelector({
  currentUser: selectCurrentUser,
  hidden: selectCartHidden
})

export default connect( mapStateToProps )(Header);
