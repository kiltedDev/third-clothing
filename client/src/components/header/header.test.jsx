import { shallow } from 'enzyme';
import React from 'react';
import { Header } from './header.component';
import { CartDropdown } from '../cart-dropdown/cart-dropdown.component';

let wrapper;
let snapshot;
let mockSignOutStart;
beforeAll(() => {
  mockSignOutStart = jest.fn();
  const mockProps = {
    currentUser: true,
    hidden: true,
    signOutStart: mockSignOutStart
  }
  wrapper = shallow(<Header { ...mockProps } />);
  snapshot = wrapper.debug();
})

describe('Header', () => {
  it('should render Header component',() => {
    expect( snapshot ).toMatchSnapshot();
  })

  describe('user is signed in', () => {
    it('should have sign out option if user is signed in',() => {
      expect(
        wrapper
        .find('OptionLink')
        .at(2)
        .text()
      ).toBe('Sign Out');
    })

    it('should call signOutStart if Sign Out is clicked',() => {
      wrapper.find('OptionLink').at(2).simulate('click')
      expect(mockSignOutStart).toHaveBeenCalled();
    })
  })

  describe('No user is signed in', () => {
    let wrapper2;
    beforeAll(() => {
      const mockProps2 = {
        currentUser: false,
        hidden: true,
        signOutStart: mockSignOutStart
      }
      wrapper2 = shallow(<Header { ...mockProps2 } />);
    })

    it('should have Sign In option if user is signed in',() => {
      expect(
        wrapper2
        .find('OptionLink')
        .at(2)
        .text()
      ).toBe('Sign In');
    })
  })

  describe('cart is hidden', () => {
    const mockProps3 = {
      currentUser: false,
      hidden: true,
      signOutStart: mockSignOutStart
    }
    const wrapper3 = shallow(<Header { ...mockProps3 } />);
    it('should not display the cart',() => {
      expect(wrapper3.exists('CartDropdown')).toBe(false);
    });
  })

  describe('cart is not hidden', () => {
    const mockProps4 = {
      currentUser: false,
      hidden: false,
      signOutStart: mockSignOutStart
    }
    const wrapper4 = shallow(<Header { ...mockProps4 } />);
    it('should display the cart',() => {
      expect(wrapper4.exists('withRouter(Connect(CartDropdown))')).toBe(true);
    })
  })
})
