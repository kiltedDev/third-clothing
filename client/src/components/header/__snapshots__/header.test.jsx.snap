// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render Header component 1`] = `
"<styled.div className=\\"header\\">
  <Styled(Link) className=\\"logo-container\\" to=\\"/\\">
    <Icon className=\\"logo\\" alt=\\"horn of odin produced by flaticon.com\\" label=\\"horn of odin produced by flaticon.com\\" title=\\"horn\\" />
  </Styled(Link)>
  <styled.div>
    <OptionLink to=\\"/shop\\" name=\\"Shop\\">
      Shop
    </OptionLink>
    <OptionLink to=\\"/contact\\" name=\\"Contact\\">
      Contact
    </OptionLink>
    <OptionLink as=\\"div\\" onClick={[Function: mockConstructor]} name=\\"Sign Out\\">
      Sign Out
    </OptionLink>
    <Connect(CartIcon) />
  </styled.div>
</styled.div>"
`;
